### ISSUE TITLE
A way to run an executable without tracking it from ad-hoc scripts

### ISSUE THREAD

**Issue by Klaim** [build2/build2#388]
Note that this feature request is specific to when using ad-hoc recipes to build convenient project-specific commands.

I have a case where in an ad-hoc script I run an executable which is an visual edition tool (here the Godot editor). That ad-hoc script will run various things to be sure everything is alright in the project and binaries are up to date before running the editor. In my case it is used by invoking `b run{editor}` (here `run` is an alias to ad-hoc target types). Running the editor is the last command in the script, and I use `env` to help control the context in which it is run.

When running that command successfully and the editor opens, which has console output, there are a few issues:
1. it's console output is buffered by `build2` so we cannot see it progressively but only once we close the editor - this can be solved by adding `--no-diag-buffer` to the `b run{editor} --no-diag-buffer`, it works although not ideal when we want it for all `run` target types, but we can ignore that;
2. the console is now waiting for `build2` + `godot.exe` to end ; we can workaround that on bash-like terminals by using `b run{editor} &` but that means that all the code before running the editor will also be in parallel, so it's not preferable.
3. I didnt try to trigger that but I suspect multiple `b` calls on the same files will cause issues, and I would like in the future to be able to have the editor (here Godot) be able to also invoke `b` (for convenient build features).

Ideally, I would like to be able to run a command but not wait for it in the ad-hoc script: if the command started, then just continue the script (and in my current case, just end normally).

I looked for ways to do this but didnt find anything so far. The usual way to do it in bash is `&` at the end of the command to run in parallel, but that doesnt work in ad-hoc scripts probably because it has already a specific meaning coming from testscript (not totally sure), for example adding it to my current line running godot:
```
E:\projects\megastructures\megastructures\tools\godot\godot.exe tools\godot{edit}
tools\buildfile:89:80: error: missing cleanup path
[...]
```

This is fine because I would prefer a more explicit and clear way to run a command separately from the current context.

Some suggestions of possible ways to do this:
- a new command just to run a command in parallel, not waiting or buffering it's result. Maybe that requires that the command is run in a separate console? I was thinking maybe equivalent to invoking a new `bash` but cross-platform and specific to `build2`?
- augment `env` with an option to do what the previous points to. After all `env` controls the running environment and it might make sense that it allows that that environment is not tracked or is run in parallel. That would also allow to have all the useful stuffs in `env` not duplicated in a separate command.

**Comment by boris-kolpackov** [comment]
A couple of thoughts:

1. Adding an option to `env` to disable `stderr` buffering sounds sensible.

2. I am not sure we want to wade into the whole job control issue (especially on Windows) that would be required if `b` were to leave some processes running.

I also don't understand what this means:

> we can workaround that on bash-like terminals by using `b run{editor}` `&` but that means that all the code before running the editor will also be in parallel, so it's not preferable.

Specifically, I don't understand what "all the code before running the editor will also be in parallel" means. Parallel with what?

What will happen in this case is the following: `b` will start and the shell, instead of waiting for it to complete, will return to the prompt (while tracking the process in the background). `b` itself will continue as before: execute the recipe commands in sequence (note: not in parallel, `&` has no effect on this) with the last one presumably starting the editor. `b` will then gets blocked waiting for the editor to exit.

To me this seems like a perfectly sensible semantics. Also, in this case, provided the editor is the last command in the recipe (so essentially the last command this `b` process will execute), there is no problem running other `b` processes on the same project.

Finally, if you really want `b run{editor}` to have the same semantics as `b run{editor}&`, I think you should be able to achieve this with something like `bash -c "editor&"`.

**Comment by Klaim** [comment]
> Specifically, I don't understand what "all the code before running the editor will also be in parallel" means. Parallel with what?

Sorry I was not very clear there: I meant that ideally I would prefer only the editor launch (the last operation) to be detached, I would prefer if the other operations before are run normally, in a blocking way. That is, `b run{editor}` would do `b install` essentially and then as the last step detach to run the editor, but wouldn't detach before that point.

> What will happen in this case is the following: [...]

Yes, while not matching exactly what I would prefer, that behavior is still good enough for my usage.
 
> To me this seems like a perfectly sensible semantics. Also, in this case, provided the editor is the last command in the recipe (so essentially the last command this `b` process will execute), there is no problem running other `b` processes on the same project.

Ok, perfect then! I was worried there would be potential corner cases on Windows. 🥲
 
> Finally, if you really want `b run{editor}` to have the same semantics as `b run{editor}&`, I think you should be able to achieve this with something like `bash -c "editor&"`.

Interesting, I'll try that, thanks!

I'm closing this issue as at this point I dont think it's worth adding anything new to `build2`, `b run{editor}&` is enough for me at the moment and I'll try that bash idea which in principle does exactly what I need.

**Comment by Klaim** [comment]
Yes it is, although as `--no-diag-buffer` helps with that, it can wait.
I'll reopen to track that.
